Allow local re-sending of txns, reset re-broadcast timer when doing so

mrmr1993 opened the pull request:
This PR changes the handling of locally-generated transactions in the transaction pool, so that
* re-sending a transaction will not fail with `already saw this`
  - this allows transactions accidentally sent out-of-order, to accounts that are yet to be created/funded, etc. to be re-issued once the problem has resolved
* re-sending transactions resets their rebroadcast timer
  - this ensures that transactions can be recovered from becoming 'stuck' in the pool by sending them again

Checklist:

- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them:

deepthiskumar commented on src/lib/network_pool/transaction_pool.ml:
```diff
@@ -1066,21 +1091,8 @@ struct
-                                Hashtbl.add_exn t.locally_generated_uncommitted
-                                  ~key:verified
-                                  ~data:(Time.now (), `Batch batch_num) ) ;
+                              if is_sender_local then
```
Should we even try to add if a locally generated command already exists? We hit the replace condition if the transaction already exists and if it is the first nonce in the pool then dependent transactions are removed. https://github.com/MinaProtocol/mina/blob/74793a26dabd381bc65d6be767fb7559e8dadf48/src/lib/network_pool/indexed_pool.ml#L889
 If it is not the first nonce, then there is an assertion failure https://github.com/MinaProtocol/mina/blob/74793a26dabd381bc65d6be767fb7559e8dadf48/src/lib/network_pool/indexed_pool.ml#L854

deepthiskumar replied:
I think we can skip re-adding locally generated txns and just update the rebroadcast timer. We may want to update the rebroadcast timer of the dependent transactions as well.

deepthiskumar replied:
I suppose we'll also want something about the transaction to change so that replacing a transaction works as expected- Maybe change `fee > replace fee` instead of `>=` to be eligible for replace?
https://github.com/MinaProtocol/mina/blob/74793a26dabd381bc65d6be767fb7559e8dadf48/src/lib/network_pool/indexed_pool.ml#L881

deepthiskumar replied:
Ah sorry, there is another duplicate check that I did not see. So we probably don't need to handle the replace case but might want to update the timer and not error out with `Duplicate` ?

mrmr1993 replied:
> Should we even try to add if a locally generated command already exists? We hit the replace condition if the transaction already exists and if it is the first nonce in the pool then dependent transactions are removed. 

And then they're re-added (at least, as far as fees will allow) shortly after, so this should be fine.
https://github.com/MinaProtocol/mina/blob/74793a26dabd381bc65d6be767fb7559e8dadf48/src/lib/network_pool/indexed_pool.ml#L913

> Ah sorry, there is another duplicate check that I did not see. So we probably don't need to handle the replace case but might want to update the timer and log a message that timer was updated in the `Duplicate` case?

Great catch, thank you! I've updated to move the logic to the `Duplicate`.

mrmr1993 commented:
Updated this with @deepthiskumar's suggestion to mark transactions as 'accepted', so they'll also be immediately rebroadcast and the command won't return an error.

I've also rebased this onto `release/1.1.6`.

deepthiskumar reviewed (approved):
looks great!
